src/models/UserRanking.js

```diff
@@ -0,0 +1,163 @@
+        }
+      },
+      nextRanking: {
+        relation: Model.HasOneRelation,
+        modelClass: UserRanking,
+        join: {
+          from: 'user_rankings.id',
+          to: 'user_rankings.prev_ranking_id'
+        }
+      }
+    }
+  }
+
+  moveAfter(prevId) {
+    // Where prevId indicates rankingA.id...
+    // rankingA --> rankingC -->   ...   --> this --> rankingD
+    //    becomes
+    // rankingA -->   this   --> rankingC --> ... --> rankingD
+
+    if (prevId === this.id) {
+      throw new Error("Cannot move ranking after itself")
+    }
+
+    if (prevId === this.prev_ranking_id) {
+      return this;
+    }
+
+    return transaction(UserRanking.knex(), async (trx) => {
+      const [rankingA] = await UserRanking.query(trx)
+        .where('id', prevId)
+        .andWhere('user_id', this.user_id)
+
+      if (prevId && !rankingA) {
+        throw new Error('No such previous ranking exists.')
+      }
+
+      const [rankingC] = await UserRanking.query(trx)
+        .where('prev_ranking_id', prevId)
+        .andWhere('user_id', this.user_id)
+
+      const [rankingD] = await UserRanking.query(trx)
+        .where('prev_ranking_id', this.id)
+        .andWhere('user_id', this.user_id)
+
+      if (rankingD) {
+        // Unplug this from list
+        await rankingD.$query(trx)
+          .patch({prev_ranking_id: this.prev_ranking_id})
+      }
+
+      if (rankingC) {
+        // Attach rankingC to this
+        await rankingC.$query(trx)
+          .patch({prev_ranking_id: this.id})
+      }
+
+      // Plug it back in at the new location
+      await this.$query(trx)
+        .patch({prev_ranking_id: prevId})
+
+      return this;
+    })
+  }
+
+  static async insertAfter(prevId, graph) {
+    // Where prevId indicates rankingA.id...
+    // rankingA --> rankingC
+    // v Becomes v
+    // rankingA --> rankingB --> rankingC
+
+    return transaction(UserRanking.knex(), async (trx) => {
+      const [rankingA] = await UserRanking.query(trx)
+        .where('id', prevId)
+        .andWhere('user_id', graph.user_id)
+
+      if (prevId && !rankingA) {
+        throw new Error('No such previous ranking exists.')
+      }
+
+      const [rankingC] = await UserRanking.query(trx)
+        .where('prev_ranking_id', prevId)
+        .andWhere('user_id', graph.user_id)
+
+      // Plug new ranking into list
+      const rankingB = await UserRanking.query(trx)
+        .insert({
+          ...graph,
+          prev_ranking_id: prevId
+        })
+
+      // Attach rankingC to this
+      if (rankingC) {
+        await rankingC.$query(trx)
+          .patch({prev_ranking_id: rankingB.id})
+      }
+
+      return rankingB
+    })
+
+    // await UserRanking.query().upsertGraph({
+    //   ...graph,
+    //   '#id': 'the-new-user_ranking'
+    //   prevRanking: {
+    //     id: prevId
+    //   },
+    //   nextRanking: {
+    //     prevRanking: {
+    //       '#ref': 'the-new-user-ranking'
+    //     }
+    //   }
+    // })
```

Comment:
A quarter to anyone that can do this with a graph upsert instead, cause that'd be cool.

Reply:
I would do it for a bob digital coin but i already own them all